Add support for SiLabs EFR32BG13P SoC

mnkp opened the pull request:
Adds support for Silicon Labs EFR32BG13P (Blue Gecko) SoC and BRD4104A (a.k.a. SLWRB4104A)
Blue Gecko Radio Board.

mnkp committed:
soc: silabs_exx32: Add support for SiLabs EFR32BG13P SoC

This commit adds support for Silicon Labs EFR32BG13P (Blue Gecko) SoC.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
boards: arm: Add support for SiLabs EFR32BG13 BRD4104A board

This commit adds support for Silicon Labs BRD4104A (a.k.a. SLWRB4104A)
Blue Gecko and BRD4250B (a.k.a. SLWRB4250B) Flex Gecko Radio Boards.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

chrta reviewed (changes requested):
Just some minor issues. As far as i can see everything else looks good.

mnkp commented on boards/arm/efr32_brd4104a/efr32_brd4104a.dts:
```diff
@@ -0,0 +1,142 @@
+		compatible = "jedec,spi-nor";
+		label = "MX25R8035F";
+		reg = <0>;
+		spi-max-frequency = <80000000>;
```
The `reg` property is required by the binding, inherited from `spi-device`.

chrta replied:
I mean the value of spi-max-frequency. This looks like 80MHz. I think 8MHz would be better.

mnkp replied:
Sorry, I was sure you're referring to reg property and only mistakenly added comment on the next line. My bad.

I checked the documentation. The `spi-max-frequency` defines "Maximum clock frequency of device's SPI interface in Hz". According to the MX25R8035F datasheet the device supports SPI clock of up to 33MHz in Ultra Low Power Mode and 108 MHz in High Performance Mode. The part mounted on BRD4104A board boots in Ultra Low Power Mode, so it supports 33MHz clock. I propose to put 33000000 as the value of the property. 

The actual frequency of the SPI bus will be much lower as the USART module on the SoC generates the SPI clock at the maximum frequency of peripheral clock rate/2, i.e. 38.4MHz/2.

chrta replied:
No problem. Yes, i think 33MHz is fine.

mnkp replied:
Fixed

mnkp commented on boards/arm/efr32_brd4104a/doc/index.rst:
```diff
@@ -0,0 +1,207 @@
+Serial Port
+===========
+
+The EFR32BG13P SoC has two USARTs and one Low Energy UARTs (LEUART).
```
Fixed

mnkp commented:
There is one more subject I would like to discuss. I failed to do it when I submitted the PR originally.

I've changed the board naming convention. We already have in tree `efr32_slwstk6061a` which is the name of the starter kit. It contains BRD4001A base board and BRD4250B radio module. Unfortunately, I couldn't follow this approach. In this case the name of the starter kit is SLWSTK6020B, however it comes with two different radio modules: BRD4100A and BRD4001A. Each has a different SoC and different flash size. So instead I adapted the name of the radio module as the name of the board. 

The advantage of this naming scheme is that it should scale. SiLabs offers a few dozen radio boards and it will be easy to add a new one if anyone requires it. The disadvantage of this naming scheme is that it will cause board name explosion. There are already plethora of entries in `boards/arm/` folder.  Another possible approach would be the one used by `boards/arm/lpcxpresso54114` where one board folder contains multiple board definitions. But this may also become unreadable. I'm not really sure what is better.

Another thing, SiLabs is using double naming scheme for their boards. BRD4104A is also known as SLWRB4104A. As an example SLWRB4104A is used on the web page listing available radio modules, however written documentation and tools typically (but not consistently) refer to the board as BRD4104A. I've chosen BRD4104A as this is the marking on the PCB. It should make it easier for the user to identify the board.

galak commented:
DEV-REVIEW: Discussed how to support multiple SoC variants on the same base board.  Suggestion is to do what we do for multiple cores with the 'board' dir being the base board and having unique defconfig/yaml/dts files for specific base board + SoC combo.  For docs if we have <base_board>_<soc>.rst for each supported combo should show the boards that are supported.

mnkp commented:
I have updated "boards: arm: Add support for SiLabs EFR32BG13 BRD4104A board" commit to follow dev-review proposal to support multiple SoC variants in the same board directory using dedicated defconfig/yaml/dts files.

PR was also rebased on the latest master, fixing some minor Kconfig issues to align it with the Kconfig clean up done in #22664. Otherwise no changes were made to this commit.

The name of the base board where the radio board can be plugged in is BRD4001A, the name of a typical radio board is BRD4104A. To avoid board names such as `efr32_brd4001a_brd4104a`, which look confusing, I decided to use the `efr32_radio` prefix for the boards. This resulted in names such as `efr32_radio_brd4104a`, `efr32_radio_brd4250b`. Comments on the naming scheme are welcome.

As a prove of concept I added also support for BRD4250B board which is already present in tree as efr32_slwstk6061a. SLWSTK6061A is the name of the starter kit which contains BRD4001A and BRD4250B boards. If this naming scheme is accepted we could deprecate efr32_slwstk6061a in the future.

I didn't review all the different datasheets, however all the SiLabs radio board and SoCs have a very similar feature set and seem to be pin compatible. Therefore most of the information related to supported features / pinout is present on the documentation page for the main board and not radio boards.

Preview of the page documenting the main board: [EFR32 Radio Boards](https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/22423/docs/boards/arm/efr32_radio/doc/index.html).

On the [Supported Boards](https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/22423/docs/boards/index.html) page links to the radio boards are inserted flat.

chrta commented on boards/arm/efr32_radio/board.cmake:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(openocd)
+board_runner_args(jlink "--device=EFR32FG1PxxxF256")
```
I believe this would have to be change depending on the daughter board and the used efr32 soc on it.

mnkp replied:
Good catch. I remembered I need to update it but then I didn't and tested with OpenOCD only. Fixed and tested with both runners.

chrta commented on boards/arm/efr32_radio/doc/index.rst:
```diff
@@ -0,0 +1,173 @@
+
+.. _efr32_radio_supported_features:
+
+Supported Features
```
It seems that the supported features are currently not the same for the efr32fg and efr32bg.

When i look at the file boards/arm/efr32_radio/efr32_radio_brd4250b.yaml it does not mention the counter, for example.

So i think either this table should be moved to the board specific sub-page or we must ensure that this table is valid for all boards, or we must include some kind of diff or exceptions on the board's sub-page.

mnkp replied:
All the SiLabs SoCs seem to be very uniform. They differ mainly in flash / RAM size and radio capabilities. I updated the .yaml file to add missing features, the tables are aligned. I propose to move data to the radio board pages if/when we find a SoC that breaks the pattern.

chrta commented on boards/arm/efr32_radio/doc/index.rst:
```diff
@@ -0,0 +1,173 @@
+.. _efr32_radio:
+
+EFR32 Radio Boards
```
Could we specifiy this a little bit more? This seems too generic for me.
If Silabs offers a new efr32 base board which provides a different pinout, we would have to add another erf32 radio board with a different name.
Could we include the base board name here (and also in the folder name)?

mnkp replied:
> The name of the base board where the radio board can be plugged in is BRD4001A, the name of a typical radio board is BRD4104A. To avoid board names such as efr32_brd4001a_brd4104a, which look confusing, I decided to use the efr32_radio prefix for the boards. This resulted in names such as efr32_radio_brd4104a, efr32_radio_brd4250b. Comments on the naming scheme are welcome.

As I explained in my comment that's something I hoped to avoid. Especially for the board name, I don't mind the title. I mean, we can have something else than `efr32_radio_` for prefix but `efr32_brd4001a_` will be tedious to use in practice.

With boards (like with samples) we don't need to keep backwards compatibility so it's not a problem to rename the boards later, if SiLabs releases a new base board.

chrta commented on boards/arm/efr32_radio/doc/brd4250b.rst:
```diff
@@ -0,0 +1,116 @@
+Overview
+********
+
+The EFR32FG1 2400/868 MHz Dual Band Radio Board is delivered as part of
```
Could be add an image of the daugher board here? This would easily show what this page is about (same case for the EFR32BG page).

chrta reviewed (changes requested):
I like how this is structured now. Only some minor comments and i have to test this on the efr32fg board a little bit.

galak commented on boards/arm/efr32_radio/board.h:
```diff
@@ -0,0 +1,14 @@
+#ifndef __INC_BOARD_H
+#define __INC_BOARD_H
+
+/* This pin is used to enable the serial port using the board controller */
```
do we really need a board.h?

mnkp replied:
I reworked the code and removed board.h.

galak commented on dts/bindings/gpio/silabs,gecko-gpio-port.yaml:
```diff
@@ -0,0 +1,19 @@
+description: SiLabs Gecko GPIO port node
+
+compatible: "silabs,gecko-gpio-port"
```
What's the plan with the family specific bindings/compatibles?

chrta replied:
See my comment above https://github.com/zephyrproject-rtos/zephyr/pull/22423#discussion_r374164670

> After this is merged, we should change the other efm32 and efr32 to use this gecko-gpio instead of efm32-gpio or efr32-gpio.

This is at least my plan :)

galak replied:
missed that, thanks

galak reviewed (commented):
In general LGTM.  A few comments/questions.

The DTS bits might need a little rework as we want to utilize a 'NODELABEL` define, but that's still in a little bit of flux right now.  More a heads up.

mnkp commented:
Updated following suggestions from @galak.
